### PR TITLE
Add documentation for Faiss byte vector

### LINKED ISSUE #7661: [DOC] Faiss Byte Vector
**What do you want to do?**
 
- [ ] Request a change to existing documentation
- [x] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request.
For lucene engine we have [Lucene byte vector feature](https://opensearch.org/docs/latest/field-types/supported-field-types/knn-vector#lucene-byte-vector), which accepts byte vectors in the range [-128 to 127] providing memory savings upto 75% when compared with fp32 vectors. But, for large scale workloads we usually prefer to use faiss engine and as of today Faiss only supports fp32 and fp16 vectors(using SQfp16). So, adding byte vector support to faiss engine helps to reduce memory requirements especially for those users who are using LLM like [Cohere Embed](https://cohere.com/blog/int8-binary-embeddings) that generates signed int8 embeddings ranging from [-128 to 127].

***Version:** List the OpenSearch version to which this issue applies, e.g. 2.14, 2.12--2.14, or all.
 2.16

**What other resources are available?** Provide links to related issues, POCs, steps for testing, etc.
https://github.com/opensearch-project/k-NN/issues/1659

### PR DESCRIPTION
**Pull request by naveentatikonda**
### Description
Add documentation for Faiss byte vector

### Issues Resolved
Closes #7661

### Version
2.17

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by heemin32** [comment]
Can we just club byte vector for both lucene and faiss as single content? From customer's perspective, there is no difference between the two.

File: _field-types/supported-field-types/knn-vector.md [hunk 1]
```diff
@@ -87,7 +87,7 @@ However, if you intend to use Painless scripting or a k-NN score script, you onl
 
 By default, k-NN vectors are `float` vectors, where each dimension is 4 bytes. If you want to save storage space, you can use `byte` vectors with the `lucene` engine. In a `byte` vector, each dimension is a signed 8-bit integer in the [-128, 127] range. 
  
-Byte vectors are supported only for the `lucene` engine. They are not supported for the `nmslib` and `faiss` engines.
+Byte vectors are supported only for the `lucene` and `faiss` engines. They are not supported for the `nmslib` engine.
```
Comment (heemin32) [inline _field-types/supported-field-types/knn-vector.md:90]: Need to change the title from `## Lucene byte` vector to `## Byte vector`

File: _field-types/supported-field-types/knn-vector.md [hunk 2]
```diff
@@ -267,6 +267,238 @@ return Byte(bval)
 ```
 {% include copy.html %}
 
+## Faiss byte vector
```
Comment (heemin32) [inline _field-types/supported-field-types/knn-vector.md:270]: We can remove entire content here and just use whatever we have for lucene byte vector. 
The content in here can be moved to tech blog if we want.

Comment (heemin32) [reply _field-types/supported-field-types/knn-vector.md:270]: As lucene does not support IVF, we can change the lucene example with faiss example including new IVF index example.

File: _search-plugins/knn/knn-vector-quantization.md [hunk 3]
```diff
@@ -13,6 +13,10 @@ By default, the k-NN plugin supports the indexing and querying of vectors of typ
 
 OpenSearch supports many varieties of quantization. In general, the level of quantization will provide a trade-off between the accuracy of the nearest neighbor search and the size of the memory footprint consumed by the vector search. The supported types include byte vectors, 16-bit scalar quantization, and product quantization (PQ).
 
+## Faiss byte vector
```
Comment (heemin32) [inline _search-plugins/knn/knn-vector-quantization.md:16]: Let's just update Lucene byte vector as Byte vector and update the content there.

**Review by heemin32 — APPROVED** [review]
LGTM

**Review by kolchfa-aws — APPROVED** [review]
LGTM

File: _field-types/supported-field-types/knn-vector.md [hunk 4]
```diff
@@ -85,23 +85,30 @@ However, if you intend to use Painless scripting or a k-NN score script, you onl
  }
  ```
 
-## Lucene byte vector
+## Byte vectors
 
-By default, k-NN vectors are `float` vectors, where each dimension is 4 bytes. If you want to save storage space, you can use `byte` vectors with the `lucene` engine. In a `byte` vector, each dimension is a signed 8-bit integer in the [-128, 127] range. 
+By default, k-NN vectors are `float` vectors, in which each dimension is 4 bytes. If you want to save storage space, you can use `byte` vectors with the `faiss` or `lucene` engine. In a `byte` vector, each dimension is a signed 8-bit integer in the [-128, 127] range.
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:90]: "each dimension consists of 4 bytes"?

File: _field-types/supported-field-types/knn-vector.md [hunk 5]
```diff
@@ -85,23 +85,30 @@ However, if you intend to use Painless scripting or a k-NN score script, you onl
...
+Byte vectors are supported only for the `lucene` and `faiss` engines. They are not supported for the `nmslib` engine.
 {: .note}
 
 In [k-NN benchmarking tests](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/perf-tool), the use of `byte` rather than `float` vectors resulted in a significant reduction in storage and memory usage as well as improved indexing throughput and reduced query latency. Additionally, precision on recall was not greatly affected (note that recall can depend on various factors, such as the [quantization technique](#quantization-techniques) and data distribution). 
 
 When using `byte` vectors, expect some loss of precision in the recall compared to using `float` vectors. Byte vectors are useful in large-scale applications and use cases that prioritize a reduced memory footprint in exchange for a minimal loss of recall.
 {: .important}
- 
+
+When using `byte` vectors with the `faiss` engine, it is recommended to use [SIMD optimization]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index#simd-optimization-for-the-faiss-engine), which helps to significantly reduce search latencies and improve indexing throughput.
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:100]:
```suggestion
When using `byte` vectors with the `faiss` engine, we recommend using [SIMD optimization]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index#simd-optimization-for-the-faiss-engine), which helps to significantly reduce search latencies and improve indexing throughput.
```

File: _field-types/supported-field-types/knn-vector.md [hunk 6]
```diff
@@ -168,6 +175,157 @@ GET test-index/_search
...
+        "dimension": 4,
+        "data_type": "byte"
+      }
+    }
+  }
+}
+```
+{% include copy-curl.html %}
+
+Ingest training data containing byte vectors into the training index:
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:200]:
```suggestion
First, ingest training data containing byte vectors into the training index:
```

File: _field-types/supported-field-types/knn-vector.md [hunk 7]
```diff
@@ -168,6 +175,157 @@ GET test-index/_search
...
+{"index": {"_index": "test-byte-ivf", "_id": "3"}}
+{"my_vector": [1, -2, 5, -50]}
+{"index": {"_index": "test-byte-ivf", "_id": "4"}}
+{"my_vector": [9, -7, 45, -78]}
+{"index": {"_index": "test-byte-ivf", "_id": "5"}}
+{"my_vector": [80, -70, 127, -128]}
+```
+{% include copy-curl.html %}
+
+Finally, search the data. Be sure to provide a byte vector in the k-NN vector field:
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:286]: Should "vector" be in code font?

File: _field-types/supported-field-types/knn-vector.md [hunk 8]
```diff
@@ -168,6 +175,157 @@ GET test-index/_search
...
+      }
+    }
+  }
+}
+```
+{% include copy-curl.html %}
+
+### Memory estimation
+
+In the best-case scenario, byte vectors require 25% of the memory that 32-bit vectors require.
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:306]:
```suggestion
In the best-case scenario, byte vectors require 25% of the memory required by 32-bit vectors.
```

File: _field-types/supported-field-types/knn-vector.md [hunk 9]
```diff
@@ -168,6 +175,157 @@ GET test-index/_search
...
+
+#### HNSW memory estimation
+
+The memory required for Hierarchical Navigable Small Worlds (HNSW) is estimated to be `1.1 * (dimension + 8 * M)` bytes/vector.
+
+As an example, assume that you have 1 million vectors with a dimension of 256 and M of 16. The memory requirement can be estimated as follows:
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:312]: "an" M?

File: _field-types/supported-field-types/knn-vector.md [hunk 10]
```diff
@@ -168,6 +175,157 @@ GET test-index/_search
...
+
+```r
+1.1 * (256 + 8 * 16) * 1,000,000 ~= 0.39 GB
+```
+
+#### IVF memory estimation
+
+The memory required for IVF is estimated to be `1.1 * ((dimension * num_vectors) + (4 * nlist * d))` bytes/vector.
+
+As an example, assume that you have 1 million vectors with a dimension of 256 and `nlist` of 128. The memory requirement can be estimated as follows:
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:322]: "an" `nlist`?

File: _ml-commons-plugin/tutorials/semantic-search-byte-vectors.md [hunk 11]
```diff
@@ -7,7 +7,7 @@ nav_order: 10
 
 # Semantic search using byte-quantized vectors
 
-This tutorial illustrates how to build a semantic search using the [Cohere Embed model](https://docs.cohere.com/reference/embed) and byte-quantized vectors. For more information about using byte-quantized vectors, see [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector/#lucene-byte-vector).
+This tutorial illustrates how to build a semantic search using the [Cohere Embed model](https://docs.cohere.com/reference/embed) and byte-quantized vectors. For more information about using byte-quantized vectors, see [Byte vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector/#byte-vectors).
```
Comment (natebower) [inline _ml-commons-plugin/tutorials/semantic-search-byte-vectors.md:10]:
```suggestion
This tutorial shows you how to build a semantic search using the [Cohere Embed model](https://docs.cohere.com/reference/embed) and byte-quantized vectors. For more information about using byte-quantized vectors, see [Byte vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector/#byte-vectors).
```

File: _search-plugins/knn/knn-index.md [hunk 12]
```diff
@@ -41,13 +41,13 @@ PUT /test-index
 ```
 {% include copy-curl.html %}
 
-## Lucene byte vector
+## Byte vectors
 
-Starting with k-NN plugin version 2.9, you can use `byte` vectors with the `lucene` engine to reduce the amount of storage space needed. For more information, see [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector).
+Starting with k-NN plugin version 2.17, you can use `byte` vectors with the `faiss` and `lucene` engine to reduce the amount of memory and storage space needed. For more information, see [Byte vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#byte-vectors).
```
Comment (natebower) [inline _search-plugins/knn/knn-index.md:46]:
```suggestion
Starting with k-NN plugin version 2.17, you can use `byte` vectors with the `faiss` and `lucene` engines to reduce the amount of required memory and storage space. For more information, see [Byte vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#byte-vectors).
```

File: _search-plugins/knn/knn-vector-quantization.md [hunk 13]
```diff
@@ -13,13 +13,13 @@ By default, the k-NN plugin supports the indexing and querying of vectors of typ
 
 OpenSearch supports many varieties of quantization. In general, the level of quantization will provide a trade-off between the accuracy of the nearest neighbor search and the size of the memory footprint consumed by the vector search. The supported types include byte vectors, 16-bit scalar quantization, and product quantization (PQ).
 
-## Lucene byte vector
+## Byte vectors
 
-Starting with k-NN plugin version 2.9, you can use `byte` vectors with the Lucene engine in order to reduce the amount of required memory. This requires quantizing the vectors outside of OpenSearch before ingesting them into an OpenSearch index. For more information, see [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector).
+Starting with version 2.17, the k-NN plugin supports `byte` vectors with the `faiss` and `lucene` engine in order to reduce the amount of required memory. This requires quantizing the vectors outside of OpenSearch before ingesting them into an OpenSearch index. For more information, see [Byte vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#byte-vectors).
```
Comment (natebower) [inline _search-plugins/knn/knn-vector-quantization.md:18]:
```suggestion
Starting with version 2.17, the k-NN plugin supports `byte` vectors with the `faiss` and `lucene` engines in order to reduce the amount of required memory. This requires quantizing the vectors outside of OpenSearch before ingesting them into an OpenSearch index. For more information, see [Byte vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#byte-vectors).
```

File: _search-plugins/knn/knn-vector-quantization.md [hunk 14]
```diff
@@ -13,13 +13,13 @@ By default, the k-NN plugin supports the indexing and querying of vectors of typ
...
+Starting with version 2.16, the k-NN plugin supports built-in scalar quantization for the Lucene engine. Unlike [byte vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#byte-vectors), which requires you to quantize vectors before ingesting the documents, the Lucene scalar quantizer quantizes input vectors in OpenSearch during ingestion. The Lucene scalar quantizer converts 32-bit floating-point input vectors into 7-bit integer vectors in each segment using the minimum and maximum quantiles computed based on the [`confidence_interval`](#confidence-interval) parameter. During search, the query vector is quantized in each segment using the segment's minimum and maximum quantiles in order to compute the distance between the query vector and the segment's quantized input vectors.
```
Comment (natebower) [inline _search-plugins/knn/knn-vector-quantization.md:22]:
```suggestion
Starting with version 2.16, the k-NN plugin supports built-in scalar quantization for the Lucene engine. Unlike [byte vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#byte-vectors), which require you to quantize vectors before ingesting documents, the Lucene scalar quantizer quantizes input vectors in OpenSearch during ingestion. The Lucene scalar quantizer converts 32-bit floating-point input vectors into 7-bit integer vectors in each segment using the minimum and maximum quantiles computed based on the [`confidence_interval`](#confidence-interval) parameter. During search, the query vector is quantized in each segment using the segment's minimum and maximum quantiles in order to compute the distance between the query vector and the segment's quantized input vectors. 
```

**Review by natebower — COMMENTED** [review]
@naveentatikonda @kolchfa-aws Please see my comments and changes and let me know if you have any questions. Thanks!

File: _field-types/supported-field-types/knn-vector.md [hunk 15]
```diff
@@ -168,6 +175,157 @@ GET test-index/_search
...
+
+### Memory estimation
+
+In the best-case scenario, byte vectors require 25% of the memory required by 32-bit vectors.
+
+#### HNSW memory estimation
+
+The memory required for Hierarchical Navigable Small Worlds (HNSW) is estimated to be `1.1 * (dimension + 8 * M)` bytes/vector.
+
+As an example, assume that you have 1 million vectors with a dimension of 256 and M of 16. The memory requirement can be estimated as follows:
```
Comment (kolchfa-aws) [inline _field-types/supported-field-types/knn-vector.md:312]:
```suggestion
As an example, assume that you have 1 million vectors with a dimension of 256 and an`m` of 16. The memory requirement can be estimated as follows:
```

File: _field-types/supported-field-types/knn-vector.md [hunk 16]
```diff
@@ -168,6 +175,157 @@ GET test-index/_search
...
+
+```r
+1.1 * (256 + 8 * 16) * 1,000,000 ~= 0.39 GB
+```
+
+#### IVF memory estimation
+
+The memory required for IVF is estimated to be `1.1 * ((dimension * num_vectors) + (4 * nlist * d))` bytes/vector.
+
+As an example, assume that you have 1 million vectors with a dimension of 256 and `nlist` of 128. The memory requirement can be estimated as follows:
```
Comment (kolchfa-aws) [inline _field-types/supported-field-types/knn-vector.md:322]:
```suggestion
As an example, assume that you have 1 million vectors with a dimension of 256 and an `nlist` of 128. The memory requirement can be estimated as follows:
```

**Pull request merged**